File: js/robots.js
function robots_initialize(backend)
{
    backend.listPorts(function(ports) {
        let options = '';

        for (let port of ports) {
            options += '<option value="'+port+'">'+port+'</option>';
        }

        $('.ports').html(options);
    });

    $('.add-robot').click(function() {
        backend.addRobot($('.ports').val());
    });

    $.get('robot.html', function(robot_template) {
        let warning = '<i class="bi text-warning bi-exclamation-circle"></i>';

        function updateInfos(div, infos) {
            let html = '';

            // Last communication message
            html += 'last message: ';
            if (infos.last_message) {
                html += Math.round(infos.last_message, 2)+'s';
            } else {
                html += 'never';
            }
            if (!infos.last_message || infos.last_message > 10) {
                html += ' '+warning;
            }
            html += '<br/>';


            // Battery level
            if ('battery' in infos.state) {
                for (let k=0; k<=1; k++) {
                    let voltage = infos.state.battery[k];
                    html += 'battery '+(k+1)+': '+voltage+'V';
                    if (voltage < 3.5) {
                        html += warning;
                    }
                    html += '<br/>';
                }
            }

            div.find('.infos').html(html);

            // Marker
            if (infos.marker) {
                $('.marker-image').html('<img src="markers/'+infos.marker+'.svg" class="m-1" width="80" />');
                $('.marker-select').val(infos.marker);
            } else {
                $('.marker-image').html('N/A');
                $('.marker-select').val('none');
            }
        }

        setInterval(function() {
            backend.getRobots(function(robots) {
                console.log(robots);
                for (let port in robots) {
                    let div = $('.robot[rel="'+port+'"]');
                    if (!div.length) {
                        let html = robot_template.replace(/{port}/g, port);
                        console.log(html);
                        $('.robots').append(html);

                        let div = $('.robot[rel="'+port+'"]');
                        div.find('.marker-select').change(function() {
                            let marker = $(this).val();
                            if (marker == 'none') {
                                marker = null;
                            }
                            backend.setMarker(port, marker);
                        });

                        div.find('.remove').click(function() {
                            backend.removeRobot(port);
                        });

                        div.find('.blink').click(function() {
                            backend.blink(port);
                        });
                    }

                    updateInfos(div, robots[port]);
                }

                $('.robot').each(function() {
                    let existingRobot = $('.robot');
                    for (let robot of existingRobot) {
                        if (!($(this).attr('rel') in robots)) {
                            $(this).remove();
                        }
                    }
                });
            });
        }, 100);

    });
}

File: app/robots.py
from serial.tools import list_ports
import time
import robot
import control
import config


class Robots:
    def __init__(self, detection):
        self.detection = detection
        self.robots = {}
        self.robots_by_marker = {}

        self.control = control.Control(self)
        self.control.start()

        for port, marker in config.config['robots']:
            self.robots[port] = robot.Robot(port, marker)
            if marker is not None:
                self.robots_by_marker[marker] = self.robots[port]

    def ports(self):
        return [entry.device for entry in list_ports.comports()]

    def saveConfig(self):
        config.config['robots'] = []
        for port in self.robots:
            config.config['robots'].append([port, self.robots[port].marker])
        config.save()
        
    def addRobot(self, port):
        if port not in self.robots:
            self.robots[port] = robot.Robot(port)
            self.saveConfig()

    def getRobots(self):
        data = {}
        for entry in self.robots:
            last_detection = None
            if self.robots[entry].marker in self.detection.last_updates:
                last_detection = time.time() - \
                    self.detection.last_updates[self.robots[entry].marker]

            data[entry] = {
                'state': self.robots[entry].state,
                'marker': self.robots[entry].marker,
                'last_detection': last_detection,
                'last_message': time.time() - self.robots[entry].last_message if self.robots[entry].last_message is not None else None
            }

        return data

    def setMarker(self, port, marker):
        if port in self.robots:
            self.robots[port].setMarker(marker)
            self.robots_by_marker[marker] = self.robots[port]
            self.saveConfig()

    def remove(self, port):
        self.robots[port].close()
        del self.robots[port]
        self.saveConfig()
